[SYCL][DOC] Introduce fuzzing test plan (#19341)
intel/llvm: sycl/doc/design/Fuzzing.md

# Fuzzing plan for intel/llvm

The main product which is being developed at intel/llvm repo is a SYCL

> implementation. At high-level, it consists of two compiler and runtime and

Reviewer: ```suggestion
implementation. At high-level it consists of two compilers and a runtime,
```

Reviewer 2: I think Alexey meant to say "two components: a compiler and a runtime" rather than "two compilers and a runtime".

therefore this document will be divided into two major sections covering those
components. Those components are essentially the only entry points through which

> a user can interact with our product.

Reviewer: It would make sense to add a short summary of what fuzzing is and why we want to do it.

## SYCL Runtime

> SYCL runtime is a library which implements SYCL API, but besides that API it

Reviewer: ```suggestion
SYCL runtime (RT) is a library which implements SYCL API, but besides that API it
```

Just because we refer to it as RT later, so it this should help the reader make the shorthand association.

Author: We only ever use `RT` abbreviation once, to I opted in for expanding it to have a clearer language in db6aba9154a2989f998a7c77ac8e388160dd9b18

also has multiple configuration options which can be tweaked through environment
variables and config file.

### Fuzzing environment variables

Following environment variables [documentation][sycl-rt-env-variables] every env
variable should be fuzzed.

> The most interesting are ones which expect data in a certain format, like

Reviewer: I would mention that we are still talking about env vars, given it is a new paragraph.

```suggestion
The most interesting of the environment variables are ones which expect data in a certain format, like
```

Author: Applied in db6aba9154a2989f998a7c77ac8e388160dd9b18

`ONEAPI_DEVICE_SELECTOR` or `SYCL_CACHE_TRESHOLD`.

[sycl-rt-env-variables]: https://github.com/intel/llvm/blob/sycl/sycl/doc/EnvironmentVariables.md

### Fuzzing sycl config file

Instead of tweaking SYCL RT behavior through environment variables, the same
can be done by providing a config file. We don't seem to have any documentation
on it except the source code for the functionality which is located in
`sycl/source/detail/config.cpp` file.

There is a prototype for the sycl config file fuzzer available at
https://github.com/intel/llvm/pull/16308

### Fuzzing API entry points

TODO: think more about this section, i.e. whether or not we want to fuzz SYCL
APIs. Not every of them accepts "raw" data, but instead expects some SYCL
objects returned from previous API calls. However, there are still plenty of
APIs which accept raw pointers and other fundamental data types. Note: to
properly fuzz them structure-aware fuzzing may be needed.

## SYCL Compiler

> SYCL compiler is based on the upstream LLVM and it is an enormously huge

Reviewer: ```suggestion
SYCL compiler is based on the [upstream LLVM compiler project](https://github.com/llvm/llvm-project) and it is an enormously huge
```

[llvm-fuzzers]: https://llvm.org/docs/FuzzingLLVM.html

However, even though we could re-use existing fuzzers, we can't just rely on
someone else running them on the upstream codebase, because those runs won't
cover any customizations we made (including new components like optimization
passes which we added only in our downstream).

There are also some unique components which may require special fuzzers.
Sections below will go through components that we have and describe in more
details like what should we fuzz and if we already have an existing fuzzer for
that.

There is also [intel/yarpgen](https://github.com/intel/yarpgen) project that can
be used to fuzz SYCL compilers. It generate random programs (of certain
structure) to detect weaknesses and bugs in optimization passes.

### Command line options

There are plenty of SYCL-specific command line options and there are multiple of
those which are not mere flag, but expect a user-provided value in a certain
format.

Those options should be fuzzed as well to ensure proper error handling of
various weird inputs.

### SYCL-specific passes

We have developed a number of passes to implement different SYCL features. They

> all can be found in `llvm/lib/SYCLLowerIR` folder. We don't need a dedicated

Author: Applied in db6aba9154a2989f998a7c77ac8e388160dd9b18

> fuzzer for every pass, we can re-use existing LLVM fuzzer intended for compiler

Reviewer: ```suggestion
fuzzer for every pass, but can instead re-use existing LLVM fuzzer intended for compiler
```

Author: Applied in db6aba9154a2989f998a7c77ac8e388160dd9b18
